deps/http.lua

```diff
@@ -363,51 +369,37 @@ function ClientRequest:initialize(options, callback)
     self.connected = true
     self:emit('socket', socket)
 
-    local function onEnd()
-      -- Just in case the stream ended and we still had an open response,
-      -- end it.
-      if res then flush() end
-    end
     local function onData(chunk)
       -- Run the chunk through the decoder by concatenating and looping
       buffer = buffer .. chunk
       while true do
         local R, event, extra = pcall(self.decode,buffer)
         if R==true then
           -- nil extra means the decoder needs more data, we're done here.
-          if not extra then break end
+          if not extra then return end
           -- Store the leftover data.
           buffer = extra
           if type(event) == "table" then
-            if self.method ~= 'CONNECT' or res == nil then
-              -- If there was an old response that never closed, end it.
-              if res then flush() end
-              -- Create a new response object
+            if not res then
+              flush()
               res = IncomingMessage:new(event, socket)
-              -- If the request upgrades the protocol then detatch the listeners so http codec is no longer used
-              local is_upgraded
-              if res.headers.upgrade then
-                is_upgraded = true
-                socket:removeListener("data", onData)
-                socket:removeListener("end", onEnd)
+              end
+            if self.method == 'CONNECT' or res.headers.upgrade then
+              local evt = self.method == 'CONNECT' and 'connect' or 'upgrade'
+              if self:listenerCount(evt) > 0 then
```

Comment:
Could you explain what this count check is for?

Reply:
See [nodejs http client source](https://github.com/nodejs/node/blob/v12.11.1/lib/_http_client.js#L479).

If you are listening on "upgrade" it means that you expect an upgrade event. If you are not listening on that event, the upgrade header should be ignored. If you are not listening on "upgrade" and the server somehow sends a 101 "switching protocols" then the connection will be destroyed.

Nodejs doc:

> Emitted each time a server responds to a request with an upgrade. If this event is not being listened for and the response status code is 101 Switching Protocols, clients receiving an upgrade header will have their connections closed.